docs/design/proposals/late-initialization/late-initialization.md

```diff
@@ -0,0 +1,113 @@
+
+This document will propose the solution for above problem and provide the details of implementing it.
+
+## Requirements
+
+A solution for this problem should :
+
+1. Update the desired state of the K8s custom resource with late initialized fields. 
+
+2. Provide configuration for AWS service teams to indicate which operation output returns the late initialized fields. i.e. Get/List/Create/Update output.
+
+    > AWS APIs have different behavior for returning the late initialized to the user. Some services return these fields as part of Create output, while some services will return the late initialized fields as part of  ReadOne output after the resource creation.
+
+3. Wait for asynchronous resource creation to populate late initialized fields.
+
+    > For some services/resources the late initialized fields do not appear immediately.  This happens when there is asynchronous processing involved in resource creation. The proposed solution should allow waiting and retry until the late initialized fields appear in the mentioned API output.
+
+4. Allow the AWS service teams to provide custom code using hooks for handing of late initialized fields.
+
+## Solution
+
+### TL;DR: 
+
+There will be a new `r.rd.AddLateInitializedFields()` method call added in the reconciliation loop, which will return `true` if the desired spec was updated with late initialized field(s) from the latest spec. If so, then reconciliation loop will patch the desired spec with new updates. 
+
+See the following sections for detailed explanation of this solution: 
+
+### Reconciler Updates
+
+* Create separate variables for `readOneLatest`, `createLatest` and `updateLatest` instead of a single `latest` variable which gets overwritten after readOne/Create/Update operation in the reconciliation loop.
+* Create a `desiredBase` which is copy of `desired` object after the status is updated(`patchResource()`) in etcd by the reconciler. We will use this copy as base to patch the desired resource spec in etcd.
+* Add `specUpdated, err := r.rd.AddLateInitializedFields(desired, readOneLatest, createLatest, updateLatest)`
+    > If the service teams have provided the configuration for populating late initialized field into `generator.yaml`, this method will update the desired spec accordingly and return `true, nil`. Otherwise it will return `false, nil`.
+* If `specUpdated` is `true`, reconciler will patch the resource in etcd using `desiredBase` as base object, otherwise no patching of K8s custom resource will take place and reconciliation will finish. 
+
+### Sdk.go Updates
+
+* Allow `sdkCreate` and `sdkUpdate` to populate the `Spec` of `latest` object and not just the `Status`.
+* This will allow capturing the late initialized fields from Create and Update API output.
+
+### ResourceDescriptor Updates
+
+* A  new method will be added in `ResourceDescriptor`, which will look like
```

Comment:
I think this functionality will be more broadly implemented than just in `ResourceDescriptor`. I think it could be extended to be a separate composition as part of `AWSResourceManager`. Further, I think it could be split into separate handlers for each operation type (`ReadOne`, `Update`, `Create`, `ReadMany`, etc.)

```golang
type AWSResourceManager interface {
    LateInitializationManager
    ...
}

type LateInitializationManager {
    SetCreateDefaults(AWSResource) (AWSResource, error)
    SetUpdateDefaults(AWSResource) (AWSResource, error)
    SetReadDefaults(AWSResource) (AWSResource, error)
    ...
}
```